pkg/yum/repository.go

```diff
@@ -422,18 +423,52 @@ func ParseRepomdXML(body io.ReadCloser) (Repomd, error) {
 }
 
 // ParseCompsXML creates PackageGroup array and Environment array from comps.xml body response
-func ParseCompsXML(body io.ReadCloser) (Comps, error) {
+func ParseCompsXML(body io.ReadCloser, url *string) (Comps, error) {
+	var reader io.Reader
 	var comps Comps
 	packageGroups := []PackageGroup{}
 	environments := []Environment{}
 
 	byteValue, err := io.ReadAll(body)
-
 	if err != nil {
 		return comps, fmt.Errorf("io.reader read failure: %w", err)
 	}
 
-	decoder := xml.NewDecoder(bytes.NewReader(byteValue))
+	fileExtension := filepath.Ext(*url)
+
+	// handle uncompressed comps
+	if fileExtension == ".xml" {
+		reader = bytes.NewReader(byteValue)
+	} else {
+		// handle compressed comps
+		bufferedReader := bufio.NewReader(bytes.NewReader(byteValue))
+
+		header, err := bufferedReader.Peek(20)
+		if err != nil {
+			return comps, err
+		}
+
+		fileType, err := filetype.Match(header)
+		if err != nil {
+			return comps, err
+		}
+
+		switch fileType {
+		case matchers.TypeGz:
+			reader, err = gzip.NewReader(bufferedReader)
+		case matchers.TypeZstd:
+			reader, err = zstd.NewReader(bufferedReader)
+		case matchers.TypeXz:
+			reader, err = xz.NewReader(bufferedReader)
+		default:
+			return comps, fmt.Errorf("invalid file type: must be gzip, xz, or zstd")
+		}
+		if err != nil {
+			return comps, fmt.Errorf("error unzipping response body: %w", err)
+		}
+	}
```

Comment:
could we combine this section with this code https://github.com/content-services/yummy/blob/master/pkg/yum/repository.go#L536-L553 to DRY it out a bit?

Reply:
done!